### ISSUE TITLE
[D-Bus API] io.crow_translate.CrowTranslate.MainWindow.translateSelection does not work in wayland.

### ISSUE THREAD

**Issue by beliys** [crow-translate/crow-translate#324]
**Bug description**
D-Bus API call `io.crow_translate.CrowTranslate.MainWindow.translateSelection` does not work in wayland (tested with KDE Shortcuts). A blank window always opens with no text selected in the browser before pressing a keyboard shortcut.

**Steps to reproduce**

1.  login to system with KDE and wayland
2.  create a keyboard shortcut as in the screenshot.
https://i.imgur.com/y3elRIt.png
3.  open any program in which you can select text (text editor, browser, etc.) and select the text
4.  press the key combination created above
5. we will see that the crow-translate window has opened without the previously selected text

**Environment**

-   OS: Manjaro
-   Desktop environment: KDE
-   Crow Translate version: 2.8.4

**Additional context**
I understand that this is most likely due to the peculiarities of the clipboard in wayland, but perhaps there is a way to make this convenient functionality work in wayland as well.

**Comment by Shatur** [comment]
Yes, as far as I know it is not possible to get the selected text in Wayland. It just doesn't have the selection buffer as in X11.

**Comment by hexauniverse** [comment]
@Shatur, in fact it's wrong. I don't know if there is any problems to get access to primary wayland selection in qt applications, but at least [wl-clipboard](https://github.com/bugaevc/wl-clipboard) has an option to use it

**Comment by Shatur** [comment]
Did not know. This is a recent addition, which is probably why it is not supported by Qt's built-in methods.
But we could use API directly. For now I'm running out of time, but if someone wants to implement it, PR is very welcome!
Should be implemented in [this](https://github.com/crow-translate/crow-translate/blob/master/src/selection.cpp). We could use `QX11Info::isPlatformX11` check to detect if we running under X11 or Wayland on Linux.

**Comment by hexauniverse** [comment]
Perhaps KDE systems require to install some additional packages (such as `wayland-protocols` for archlinux, which is required by gtk, so I have no problems in my gnome session)

That's something more i've found https://wayland.app/protocols/primary-selection-unstable-v1

**Comment by VolkMilit** [comment]
> That's something more i've found https://wayland.app/protocols/primary-selection-unstable-v1

Yes, I think it [supported by Qt](https://doc.qt.io/qt-5/qtwaylandcompositor-attribution-wayland-primary-selection-protocol.html).

**Comment by rizzini** [comment]
Any update on this?

**Comment by Shatur** [comment]
Do not bump the issue without any significant information.

**Comment by rizzini** [comment]
One of the key features is not working with the upcoming server display, which many distros already made the default, the issue doesn't get any attention whatsoever in five months and that's your response? Well.. Thanks for making things really easy for me. I'm ditching crow and I'll not recommend it anymore. In fact, I'll save this issue for further references. Take care.

**Comment by Shatur** [comment]
> One of the key features is not working with the upcoming server display, which many distros already made the default, the issue doesn't get any attention whatsoever in five months and that's your response?

I can't work on all submitted issues myself. This is not how open source works. I don't use Wayland and it's not even installed on my system. If it doesn't work in your setup - try to help improve it by digging into the problem / submitting a PR instead just bumping the issue and insulting the developer. I provided enough information about implementation details and I know that the problem is important.

> Is it just not possible?

Pretty sure it is possible. You added this part of the message later, so I didn't see it.

**Comment by VolkMilit** [comment]
> One of the key features is not working with the upcoming server display, which many distros already made the default, the issue doesn't get any attention whatsoever in five months and that's your response? Well.. Thanks for making things really easy for me. I'm ditching crow and I'll not recommend it anymore. In fact, I'll save this issue for further references. Take care.

It's opensource, go fork it, or go help to it. Whatever. And if you doesn't recommended it means nothing to opensource software, because closesource can came to you and said: "Hey! We revoke your license. Have fan with broken or non-working shitware, you can't even compile yourself!"

Now, about Wayland part. Wayland is "coming" to Linux for about 10 years now, and still nothing has moved off the dead center. Only RedHat (Fedora - also RedHat; Ubuntu is "partners", whatever) with their GNOME (yes, GNOME and GTK is RedHat toy) really support Wayland. Nvidia support *some* Wayland's composers. AMD supporting all of them in opensource drivers, but with frequent problems, like full desktop crash. KDE on other hand still doesn't fully support Wayland. WINE still doesn't support Wayland properly. Half of the Linux software doesn't support it. So, there will be Wayland, but not in this decade.

**Comment by beliys** [comment]
Perhaps this merge will be useful for the implementation of this functionality - https://invent.kde.org/plasma/kwin/-/merge_requests/1272/diffs#0aabb9eb08e49161f3f0b57ed24b7da31c19da0d

**Comment by VolkMilit** [comment]
On Arch Linux, yes, maybe. Not on something like Debian or Ubuntu.

**Comment by VolkMilit** [comment]
> > On Arch Linux, yes, maybe. Not on something like Debian or Ubuntu.
> 
> Why were these distributions mentioned? When I posted the link to the merge, I thought that this is a sample code for working "primary selection" through KWaylandServer, which could be useful.

Because this version of KWin not yet on Debian 10. Stable version of this distro still use 5.20.

**Comment by VolkMilit** [comment]
I still don't understand Wayland. It is so fragmented. Is there even API for "rule for them all", or every DE pulls the blanket over themselfs? I don't think we can support every DE and Wayland composer possible, if there is no DE-agnostic API.

**Comment by Shatur** [comment]
I still think it is possible, [wl-clipboard](https://github.com/bugaevc/wl-clipboard) somehow works on any DE.

**Comment by gzqx** [comment]
*I am not very familiar with the topic so my words might be very wrong*

I think wl-clipboard needs user to actively copy it. I was using goldendict, whose popup function also don't work in wayland. I deal with it by using setting a shortcut <X> for passing `$(wl-paste)` to goldendict. Then I do <C-C><X> to translate selection as a workaround.

It seems that `crow` will launch the GUI interface, but `crow "text"` will output result in CLI. Adding an option to launch GUI with given text will allow above workaround to be used in crow.

**Comment by Shatur** [comment]
Yes, this will work.
But I would prefer to have a proper integration. I just currently don't have time for this and not use Wayland. If someone want to tackle this - I can guide through the code.

**Comment by VolkMilit** [comment]
It's 2023 and Waylands STILL doesn't drop-in replacement for X11. Even if you wish to tinker with it, most things that works in X11, doesn't in Wayland. I'm not hating Wayland in this post and I'm not whining. In fact, I want Wayland to be standard soon enough, because my main OS for gaming is Linux and Wayland looks very smooth. But it just doesn't work and tends to break things. I really want starting to integrate Wayland in Crow, but I'm using KDE and doesn't want to switch to GNOME or some marginal WMs.

Anybody here use(d) KDE + Wayland? Is it still that bad, as in early 2022? Just curios.

**Comment by gzqx** [comment]
@VolkMilit I have used KDE+Wayland for 2 years. I use KDE+Wayland on my laptop and KDE+X11 on my desktop so I can provide some comparison.
TLDR: At least 95% of problem are resolved now. 

A year ago there are still quite some broken features, currently the thing that does not work in my mind only include:
- get selected text, which is the problem Crow, Goldendict etc face.
- Barrier, a software used to share mouse and keyboard input across devices (as far as I know they have started wayland integration)
- Software using very old electron version may have scaling problem on high res screen.

However, notice that I am using Fedora and closely follow its update, thus my package and kernel are usually a lot version newer than most distributions. I am now sure if distributions like Debian are in this state as well.

**Comment by graincoffee** [comment]
Hello guys, in my system (archlinux + kde + wayland) I found a way to achieve this function: press the shortcut key twice in succession. This translates the selected text, although the window does not appear at the cursor position .

**Comment by voltkde** [comment]
Hello guys, in my system (archlinux + kde + wayland) I found a way to achieve this function: press the shortcut key (ctrl + alt + e) ​​twice in succession. This can translate the selected text, although The window does not appear at the cursor position.

**Comment by VolkMilit** [comment]
@voltkde, can you... Uh... explain? So, if you're press shortcut once, nothing happens? And in what apps this is working? I have strong feeling, that you're using your app through XWayland, because we doesn't have api for working with selection in Wayland in Crow, as Shatur already said in first comment.

**Comment by voltkde** [comment]
> @voltkde, can you... Uh... explain? So, if you're press shortcut once, nothing happens? And in what apps this is working? I have strong feeling, that you're using your app through XWayland, because we doesn't have api for working with selection in Wayland in Crow, as Shatur already said in first comment.

When I press the shortcut key once, the translation window appears, but there is no content. 
Press the shortcut key again, and the second translation window will appear while the first window disappears. 
The second translation window will have the translated text.
I tested it with xeyes, the translation window does not use xwayland

In fact,I cloned the source code and set a breakpoint in the "void SourceTextEdit::replaceText(const QString &text)" function of the sourcetextedit.cpp file. 
When the shortcut key is pressed for the second time, the value of the variable "text" is not empty and It is indeed the text I selected (the value of the text variable is empty when the shortcut key is pressed for the first time), 
which means that when the shortcut key is pressed for the second time, the code "emit requestedSelectionAvailable(QGuiApplication::clipboard()->text(QClipboard::Selection));" (the source file is selection.cpp, 
the function is void Selection::getSelection()) worked correctly

The expression may not be clear, please forgive me :)

**Comment by VolkMilit** [comment]
> When I press the shortcut key once, the translation window appears, but there is no content.
Press the shortcut key again, and the second translation window will appear while the first window disappears.

Odd. But this bug also happens on X11 sometimes. It's too random to debug.

> I tested it with xeyes, the translation window does not use xwayland

Indeed, Crow is using Wayland directly, that what I mean. But xeyes using XWayland.

> In fact,I cloned the source code and set a breakpoint in the "void SourceTextEdit::replaceText(const QString &text)" function of the sourcetextedit.cpp file.

Thanks for debugging! But we still need to implement proper Wayland selection protocol for Crow, which is non-existed right now.

**Comment by AbaoFromCUG** [comment]
My Env

* Hyprland(Wayland) 
* wl-clipboard
* qt5-wayland
* crow-translate

Same issue happend, So I have write a simple qt5 demo:

```cpp
int main(int argc, char *argv[]) {
    QApplication a(argc, argv);
    // select me
    auto env = QProcessEnvironment::systemEnvironment();
    qDebug() << "env:QT_QPA_PLATFORM=" << env.value("QT_QPA_PLATFORM");
    qDebug() << "QApplication::platformName= " << a.platformName();
    QClipboard *clipboard = QApplication::clipboard();
    qDebug() << "QApplication::clipboard()->text(): " << clipboard->text();
    qDebug() << "QApplication::clipboard()->text(QClipboard::Selection): " << clipboard->text(QClipboard::Selection);
    qDebug() << "QApplication::clipboard()->text(QClipboard::Clipboard): " << clipboard->text(QClipboard::Clipboard);
    qDebug() << "QApplication::clipboard()->text(QClipboard::FindBuffer): " << clipboard->text(QClipboard::FindBuffer);

    QProcess *process = new QProcess(&a);
    process->start("wl-paste", {"--primary"});
    QObject::connect(process, &QProcess::readyRead, [process, &a]() {
        qDebug() << "wl-paste --primary" << process->readAllStandardOutput();
        process->close();
        a.quit();
    });
    a.setQuitOnLastWindowClosed(false);
    return QApplication::exec();
}

```

here is the output, when I select the `select me` 

```txt
/home/abao/Documents/demo/qselection_demo/cmake-build-debug/qselection_demo -platform wayland
env:QT_QPA_PLATFORM= "wayland"
QApplication::platformName=  "wayland"
QApplication::clipboard()->text():  ""
QApplication::clipboard()->text(QClipboard::Selection):  ""
QApplication::clipboard()->text(QClipboard::Clipboard):  ""
QApplication::clipboard()->text(QClipboard::FindBuffer):  ""
wl-paste --primary "select me"
```

I guess the cause of this issue is [Qt Wayland platform plugin](https://github.com/qt/qtwayland) implement issue. 

A solution: if `platform` is wayland, use QProcess to get selection content.  Is it conform to code of condust ?

**Comment by Shatur** [comment]
> A solution: if platform is wayland, use QProcess to get selection content.

Maybe we could just get the code from `wl-paste` and do the same? It will be faster then launching a separate process.

**Comment by AbaoFromCUG** [comment]
> Maybe we could just get the code from `wl-paste` and do the same?

primary selection is a unstable api, the `wl-paste` include three `Wayland Primary selection`  implement backend

* [primary-selection-unstable-v1](https://wayland.app/protocols/primary-selection-unstable-v1)
* [wlr-data-control-unstable-v1 (wlroots)](https://wayland.app/protocols/wlr-data-control-unstable-v1)
* gtk-primary-selection implementation

unfortunatily, there isn't libary named `libwl-clipboard` :cry:

**Comment by gzqx** [comment]
I have been using wl-paste for over a year in my workaround and it seems working without problem. Well of course, if what you worried about is future changes can break the function then I agree it is possible.

**Comment by Shatur** [comment]
No, no, I just considering "native" way first. If there is a lot of code to copy - it's better to use `QProcess`.

**Comment by AbaoFromCUG** [comment]
QtWayland module wrap the wayland protocol.  As a wayland client, Qt wrap wayland event loop, and implement lots of `QxxxPlatform`  interface.

On the one hand, we can't get [wayland global variable](https://bugaevc.gitbooks.io/writing-wayland-clients/content/black-square/global-objects.html) which needed to obtain `primary_selection_device_manager`  via raw wayland protocol. Regarding to https://bugaevc.gitbooks.io/writing-wayland-clients/content/black-square/global-objects.html

On the other hand, Qt native interface is private header,  It isn't convenient to obtain raw `wayland global variable` via Qt native interface.  More important, the private header will change in future

Some native interface map: 
- [QWaylandDislay](https://bugaevc.gitbooks.io/writing-wayland-clients/content/black-square/global-objects.html) - wl_display
* [QWaylandPrimarySelectionDeviceManagerV1](https://github.com/qt/qtwayland/blob/5.15/src/client/qwaylandprimaryselectionv1.cpp#L53) - [zwp_primary_selection_device_manager_v1]([zwp_primary_selection_device_manager_v1](https://wayland.app/protocols/primary-selection-unstable-v1#zwp_primary_selection_device_manager_v1))
* [QWaylandPrimarySelectionSourceV1](https://github.com/qt/qtwayland/blob/5.15/src/client/qwaylandprimaryselectionv1.cpp#L136)  - [zwp_primary_selection_source_v1](https://wayland.app/protocols/primary-selection-unstable-v1#zwp_primary_selection_source_v1)

**Comment by Shatur** [comment]
Okay, let's go with `QProcess` then until `QtWayland` missing this feature.

**Comment by frijheid** [comment]
I resolved the issue by adding Crow Translate to the system's shortcut settings, enabling me to assign keys for available commands. Users can try adding it to their shortcut settings to potentially resolve the issue.
OS: Fedora KDE
